src/child.rs

```diff
@@ -0,0 +1,176 @@
+}
+
+fn print_child_output<S>(out: Result<Output<S>, io::Error>, command_name: &str)
+where
+    S: AsRef<str>,
+{
+    let message = match out {
+        Ok(Output::Stdout(line)) => format!("{} (stdout): {}", command_name, line.as_ref()),
+        Ok(Output::Stderr(line)) => format!("{} (stderr): {}", command_name, line.as_ref()),
+        Err(e) => format!("error reading {} output: {}", command_name, e),
+    };
+    PBAR.message(&message);
+}
+
+fn handle_output<I, S>(
+    output: I,
+    logger: &Logger,
+    should_print: bool,
+    stdout: &mut String,
+    stderr: &mut String,
+    command_name: &str,
+) where
+    I: IntoIterator<Item = Result<Output<S>, io::Error>>,
+    S: AsRef<str>,
+{
+    for out in output {
+        match out {
+            Ok(Output::Stdout(ref line)) => {
+                let line = line.as_ref().trim_end();
+                info!(logger, "{} (stdout): {}", command_name, line);
+                stdout.push_str(line);
+            }
+            Ok(Output::Stderr(ref line)) => {
+                let line = line.as_ref().trim_end();
+                info!(logger, "{} (stderr): {}", command_name, line);
+                stderr.push_str(line);
+            }
+            Err(ref e) => {
+                warn!(logger, "error reading output: {}", e);
+            }
+        }
+        if should_print {
+            print_child_output(out, command_name);
+        }
+    }
+}
+
+/// Run the given command and return its stdout.
+///
+/// If the command takes "too long", then its stdout and stderr are also piped
+/// to our stdout and stderr so that the user has an idea of what is going on
+/// behind the scenes.
+pub fn run(
+    logger: &Logger,
+    mut command: process::Command,
+    command_name: &str,
+) -> Result<String, failure::Error> {
+    info!(logger, "Running {:?}", command);
+
+    let mut child = command
+        .stdout(process::Stdio::piped())
+        .stderr(process::Stdio::piped())
+        .spawn()?;
+
+    let since = time::Instant::now();
+
+    let stdout = io::BufReader::new(child.stdout.take().unwrap());
+    let stderr = io::BufReader::new(child.stderr.take().unwrap());
+
+    let (send, recv) = mpsc::channel();
+    let stdout_send = send.clone();
+    let stderr_send = send;
+
+    // Because pipes have a fixed-size buffer, we need to keep reading stdout
+    // and stderr on a separate thread to avoid potential dead locks with
+    // waiting on the child process.
+
+    let stdout_handle = thread::spawn(move || {
+        for line in stdout.lines() {
+            stdout_send.send(line.map(Output::Stdout)).unwrap();
+        }
+    });
+
+    let stderr_handle = thread::spawn(move || {
+        for line in stderr.lines() {
+            stderr_send.send(line.map(Output::Stderr)).unwrap();
+        }
+    });
+
+    let mut stdout = String::new();
+    let mut stderr = String::new();
+    let mut is_long_running = false;
+
+    loop {
+        if !is_long_running && taking_too_long(since) {
```

Comment:
I'm not sure I'm 100% convinced this strategy of swallowing the first 200ms of output will work out in the long run. In the worst case a tool finishes very quickly but some crucial information it prints to stdout/stderr is lost. I don't think that wasm-pack invokes many tools like that today though? This is largely just in terms of future concerns.

In general I feel we should depend on tools responsibly not generating lots of output that's not necessary. Like adding a wasm target with rustup which is already there should produce zero output, but I'm not sure that we should gauge that based on the 200ms delay here. (similarly Cargo's "fresh" output of "you compiled this in 0.2s" is something we'd want to hide, but there isn't a great way to do that in Cargo right now unfortunately).

Reply:
So you're saying you think we should always display the child's output unconditionally?

Reply:
After talking over lunch I think the conclusion was that we should probably remove this yeah, and have the main thread block on dispatching output received from the channel

Reply:
For context:

* We *really* don't want to swallow warnings from rustc, even when cargo build exits super quickly.

* That general scenario could play out with other CLI tools too

* The noisiest parts of the `wasm-pack build` output are actually things like "checking if you have wasm32-u-u installed" when 99% of the time the user already does. For this particular case, it is wasm-pack itself printing this out, not a child process, so this module is unaffected.

Therefore, we should always print all output, regardless of time the child takes to run.

For that last bullet point, and any kind of "install X if it isn't already here", we only want to print in the case where X isn't already installed and we are going to commence with installing it.

Reply:
Er and to elaborate a bit! I think it's fine to leave this in if it's just hypothetical things I'm worried about and I can always file a follow-up issue, but one concrete use case we discovered which may turn out badly is a `rustc` invocation which prints a compiler warning (like a diagnostic) but finishes in under 200ms. I think the logic here would swallow up the warning and avoid it getting printed.

Reply:
Er oops sorry, double post at the same time!